File: jgraph/personal-email-exporter/src/main/java/com/microsoft/graphdataconnect/skillsfinder/personal/email/exporter/services/GraphService.java
package com.microsoft.graphdataconnect.skillsfinder.personal.email.exporter.services;

import com.microsoft.graph.logger.DefaultLogger;
import com.microsoft.graph.logger.LoggerLevel;
import com.microsoft.graph.models.extensions.IGraphServiceClient;
import com.microsoft.graph.models.extensions.MailFolder;
import com.microsoft.graph.models.extensions.User;
import com.microsoft.graph.options.Option;
import com.microsoft.graph.options.QueryOption;
import com.microsoft.graph.requests.extensions.GraphServiceClient;
import com.microsoft.graph.requests.extensions.IMessageCollectionPage;
import com.microsoft.graph.requests.extensions.IUserCollectionPage;

import java.io.BufferedInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.LinkedList;
import java.util.List;

public class GraphService {

    public static final int HARDCODED_EMAILS_PER_PAGE = 20;
    public static final int HARDCODED_USERS_PER_PAGE = 20;

    private final String accessToken;

    private IGraphServiceClient graphClient = null;
    private SimpleAuthProvider authProvider = null;

    public GraphService(String accessToken) {
        this.accessToken = accessToken;
    }

    private void ensureGraphClient(String accessToken) {
        if (graphClient == null) {
            // Create the auth provider
            authProvider = new SimpleAuthProvider(accessToken);

            // Create default logger to only log errors
            DefaultLogger logger = new DefaultLogger();
            logger.setLoggingLevel(LoggerLevel.ERROR);

            // Build a Graph client
            graphClient = GraphServiceClient.builder()
                    .authenticationProvider(authProvider)
                    .logger(logger)
                    .buildClient();
        }
    }

    public User getUser() {
        ensureGraphClient(accessToken);

        // GET /me to get authenticated user
        User me = graphClient
                .me()
                .buildRequest()
                .get();

        return me;
    }

    public List<MailFolder> getMailDirectories() {
        ensureGraphClient(accessToken);
        final List<Option> options = new LinkedList<>();
        options.add(new QueryOption("top", "300"));
        return graphClient.me().mailFolders().buildRequest(options).get().getCurrentPage();
    }

    public IMessageCollectionPage getMailsFromFolder(String mailFolderId, String columnToFilter) {
        ensureGraphClient(accessToken);
        // receivedDateTime
        final List<Option> options = new LinkedList<>();
        options.add(new QueryOption("orderby", columnToFilter + " DESC"));
        options.add(new QueryOption("top", String.valueOf(HARDCODED_EMAILS_PER_PAGE)));
        return graphClient.me().mailFolders(mailFolderId).messages().buildRequest(options).get();
    }

    public IUserCollectionPage readUsers() {
        ensureGraphClient(accessToken);
        return graphClient.users().buildRequest().get();
    }

    public byte[] readUserProfilePicture(String mail) {
        ensureGraphClient(accessToken);
        try {
            BufferedInputStream stream = graphClient.customRequest("/users/" + mail + "/photo/$value", BufferedInputStream.class)
                    .buildRequest()
                    .get();

            return readStream(stream);
        } catch (Exception e) {
            return null;
        }
    }

    private byte[] readStream(BufferedInputStream stream) {
        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
        int nRead;
        byte[] data = new byte[16384];

        while (true) {
            try {
                if ((nRead = stream.read(data, 0, data.length)) == -1) break;
                buffer.write(data, 0, nRead);
            } catch (IOException e) {
                e.printStackTrace();
            }

        }

        return buffer.toByteArray();

    }

}

File: solutions/projectstaffing/jgraph/m365-user-to-azure-sql/src/main/scala/com/microsoft/graphdataconnect/skillsfinder/userdetails/job/service/GraphService.scala
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project root for full license information.
 */

package com.microsoft.graphdataconnect.skillsfinder.userdetails.job.service

import java.io.BufferedInputStream

import com.microsoft.graph.authentication.IAuthenticationProvider
import com.microsoft.graph.http.IHttpRequest
import com.microsoft.graph.logger.LoggerLevel
import com.microsoft.graph.models.extensions._
import com.microsoft.graph.requests.extensions.GraphServiceClient
import com.microsoft.graphdataconnect.logging.GdcLogger

class GraphService(graphClient: IGraphServiceClient, log: GdcLogger) {
  // According to https://docs.microsoft.com/en-us/graph/api/profilephoto-get?view=graph-rest-1.0
  // "The supported sizes of HD photos on Microsoft 365 are as follows: 48x48, 64x64, 96x96, 120x120, 240x240, 360x360,
  // 432x432, 504x504, and 648x648. Photos can be any dimension if they are stored in Azure Active Directory."
  private val DEFAULT_MAX_PICTURE_SIZE: String = "120x120"

  def readUserProfilePicture(mail: String): Option[Array[Byte]] = {
    val sizedProfilePicture: Option[Array[Byte]] = getSizedProfilePicture(mail)
    if (sizedProfilePicture.isDefined) {
      sizedProfilePicture
    } else {
      getDefaultSizedProfilePicture(mail)
    }
  }

  private def getDefaultSizedProfilePicture(mail: String): Option[Array[Byte]] = {
    try {
      val stream = graphClient.customRequest("/users/" + mail + "/photo/$value", classOf[BufferedInputStream]).buildRequest.get
      Some(Stream.continually(stream.read).takeWhile(_ != -1).map(_.toByte).toArray)
    } catch {
      case e: Exception =>
        log.error(s"Cannot read default profile picture for user with mail: $mail.", e)
        None
    }
  }

  private def getSizedProfilePicture(mail: String, profilePictureSize: String = DEFAULT_MAX_PICTURE_SIZE): Option[Array[Byte]] = {
    try {
      val stream = graphClient.customRequest(s"/users/$mail/photos/$profilePictureSize/" + "$value", classOf[BufferedInputStream]).buildRequest.get
      Some(Stream.continually(stream.read).takeWhile(_ != -1).map(_.toByte).toArray)
    } catch {
      case e: Exception =>
        log.error(s"Cannot read $profilePictureSize profile picture for user with mail: $mail. Trying to get the default picture", e)
        None
    }
  }


}

object GraphService {

  def apply(accessToken: String, iLogger: GdcILogger): GraphService = new GraphService(initGraphClient(accessToken, iLogger), iLogger.gdcLogger)

  private def initGraphClient(accessToken: String, iLogger: GdcILogger): IGraphServiceClient = {
    val authProvider = new SimpleAuthProvider(accessToken)
    iLogger.setLoggingLevel(LoggerLevel.ERROR)
    // Build a Graph client
    GraphServiceClient.builder.authenticationProvider(authProvider).logger(iLogger).buildClient
  }

  private class SimpleAuthProvider(val accessToken: String) extends IAuthenticationProvider {
    override def authenticateRequest(request: IHttpRequest): Unit = {
      // Add the access token in the Authorization header
      request.addHeader("Authorization", "Bearer " + accessToken)
    }
  }

}